pkg/registration/hub/importer/importer.go

```diff
@@ -0,0 +1,284 @@
+	newCluster, err = i.reconcile(ctx, logger, syncCtx.Recorder(), provider, newCluster)
+	updated, updatedErr := i.patcher.PatchStatus(ctx, newCluster, newCluster.Status, cluster.Status)
+	if updatedErr != nil {
+		return updatedErr
+	}
+	if err != nil {
+		return err
+	}
+	if updated {
+		syncCtx.Recorder().Eventf(
+			"ManagedClusterImported", "managed cluster %s is imported", clusterName)
+	}
+
+	return nil
+}
+
+func (i *Importer) reconcile(
+	ctx context.Context,
+	logger klog.Logger,
+	recorder events.Recorder,
+	provider cloudproviders.Interface,
+	cluster *v1.ManagedCluster) (*v1.ManagedCluster, error) {
+	clients, err := provider.Clients(ctx, cluster)
+	if err != nil {
+		meta.SetStatusCondition(&cluster.Status.Conditions, metav1.Condition{
+			Type:   ConditionTypeImported,
+			Status: metav1.ConditionFalse,
+			Reason: "KubeConfigGetFailed",
+			Message: fmt.Sprintf("failed to get kubeconfig. See errors:\n%s",
+				err.Error()),
+		})
+		return cluster, err
+	}
+
+	if clients == nil {
+		meta.SetStatusCondition(&cluster.Status.Conditions, metav1.Condition{
+			Type:    ConditionTypeImported,
+			Status:  metav1.ConditionFalse,
+			Reason:  "KubeConfigNotFound",
+			Message: "Secret for kubeconfig is not found.",
+		})
+		return cluster, nil
+	}
+
+	// render the klsuterlet chart config
+	klusterletChartConfig := &chart.KlusterletChartConfig{
+		CreateNamespace: true,
+		Klusterlet: chart.KlusterletConfig{
+			Create:      true,
+			ClusterName: cluster.Name,
+			ResourceRequirement: operatorv1.ResourceRequirement{
+				Type: operatorv1.ResourceQosClassDefault,
+			},
+		},
+	}
+	for _, renderer := range i.renders {
+		klusterletChartConfig, err = renderer(ctx, klusterletChartConfig)
+		if err != nil {
+			meta.SetStatusCondition(&cluster.Status.Conditions, metav1.Condition{
+				Type:   ConditionTypeImported,
+				Status: metav1.ConditionFalse,
+				Reason: "ConfigRendererFailed",
+				Message: fmt.Sprintf("failed to render config. See errors:\n%s",
+					err.Error()),
+			})
+			return cluster, err
+		}
+	}
+	rawManifests, err := chart.RenderKlusterletChart(klusterletChartConfig, operatorNamesapce)
+	if err != nil {
+		return cluster, err
+	}
+
+	clientHolder := resourceapply.NewKubeClientHolder(clients.KubeClient).
+		WithAPIExtensionsClient(clients.APIExtClient).WithDynamicClient(clients.DynamicClient)
+	cache := resourceapply.NewResourceCache()
+	var results []resourceapply.ApplyResult
+	for _, manifest := range rawManifests {
+		requiredObj, _, err := genericCodec.Decode(manifest, nil, nil)
+		if err != nil {
+			logger.Error(err, "failed to decode manifest", "manifest", manifest)
+			return cluster, err
+		}
+		result := resourceapply.ApplyResult{}
+		switch t := requiredObj.(type) {
+		case *appsv1.Deployment:
+			result.Result, result.Changed, result.Error = resourceapply.ApplyDeployment(
+				ctx, clients.KubeClient.AppsV1(), recorder, t, 0)
+			results = append(results, result)
+		case *operatorv1.Klusterlet:
```

Comment:
Here I don't quite understand why we handle the Deployment and Klusterlet separately, instead of using the resourceapply.ApplyDirectly?

Reply:
it only support ApplyDirectly only can apply certain static resource, deployment/klusterlet can do be applied with this func.